src/Microsoft.TemplateEngine.TemplateLocalizer/Commands/Export/ExportCommand.cs

```diff
@@ -0,0 +1,196 @@
+                {
+                    // No new files has been added by this provider. This is an indication of a bad input.
+                    Logger.LogError(LocalizableStrings.command_export_log_templateJsonNotFound, templateJsonProvider.Path);
+                    failed = true;
+                }
+            }
+
+            if (failed)
+            {
+                Logger.LogError(LocalizableStrings.generic_log_commandExecutionFailed, CommandName);
+                return 1;
+            }
+
+            ExportOptions exportOptions = new ExportOptions()
+            {
+                Languages = args.Languages,
+                DryRun = args.DryRun,
+            };
+
+            List<ExportResult> exportResults = new List<ExportResult>();
+            List<Task<ExportResult>> runningExportTasks = new List<Task<ExportResult>>(templateJsonFiles.Count);
+            foreach (FileInfo templateJsonFile in templateJsonFiles)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    break;
+                }
+
+                string templateDirectory = Path.GetDirectoryName(templateJsonFile.FullName) ?? string.Empty;
+                exportOptions.TargetDirectory = Path.Combine(templateDirectory, "localize");
+
+                runningExportTasks.Add(new Core.TemplateLocalizer(Logger).ExportLocalizationFiles(templateJsonFile.FullName, exportOptions, cancellationToken));
+
+                if (runningExportTasks.Count == ConcurrencyLevel)
+                {
+                    // We have reached the concurrency limit. Wait for one of the tasks to finish before issuing new ones.
+                    Task<ExportResult>? completedTask = null;
+
+                    try
+                    {
+                        completedTask = await Task.WhenAny(runningExportTasks).ConfigureAwait(false);
+                        ExportResult taskResult = await completedTask.ConfigureAwait(false);
+                        exportResults.Add(taskResult);
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        // Task was intentionally cancelled.
+                        Logger.LogInformation(LocalizableStrings.command_export_log_cancelled);
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.LogError(LocalizableStrings.generic_log_commandExecutionFailedWithErrorMessage, e.Message);
+                        // Fatal error. Stop processing the rest of the files.
+                        cts.Cancel();
+                        failed = true;
+                        break;
+                    }
+
+                    if (completedTask != null)
+                    {
+                        runningExportTasks.Remove(completedTask);
+                    }
+                }
+            }
+
+            try
```

Comment:
please modularize this method, it will make life easier moving forward

Reply:
+1. Also, writing what's basically a scheduler with a given max concurrency level feels like a problem that must have been solved elsewhere. What would be the downside of doing this with a `Parallel.ForEach` for example?

Reply:
Also, what is the downside of just let TPL schedule this work here? In most of the cases we would have just one task. Is implementing the scheduler worth it? I feel like reaching out 16 is 0.01% case. 
I see the point in such implementation when it's likely to produce lots of tasks which may flood or you want to limit it more than max amount of threads. In case control is needed, I would consider using `Parallel` as well.

Reply:
`Parallel.ForEach` is a blocking call and unfortunately doesn't support async delegates. This probably won't surface as a big issue since this is part of the console app and not the API, but this was the original idea I had behind staying away from `Parallel.` here.

The only replacements I could find that fully supports `async/await`:
- Using semaphore slim as [suggested here](https://devblogs.microsoft.com/pfxteam/implementing-a-simple-foreachasync/)
- Creating a scheduler as done in this example in docs page which I wish was built into the libraries: [LimitedConcurrencyLevelTaskScheduler](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler?view=net-5.0#examples)

The official support for [Parallel.ForEachAsync](https://github.com/dotnet/runtime/pull/46943) is here, but available starting .NET 6.0.

Any ideas @ladipro @vlada-shubina @donJoseLuis ?

Reply:
I would simply use `await Task.WhenAll(exportTasks).ConfigureAwait(false);` here, as mentioned before I don't see use case for slicing tasks here, and reaching out 16 concurrent tasks is very little chance.

Reply:
16 is an arbitrary number I picked here, but the number of tasks going beyond 16 is not hard at all. We already have 6 templates in our repo under ProjectTemplates.6.0 folder. There will be 6 more with .NET 7. This is just for our very basic templates and I'm not considering ItemTemplates here. I imagine this may grow significantly for a repo where templates are more important, such as aspnetcore.

I don't believe we will re-invoke the tool for every template.json in our repo. This will be too much to manage from the build pipeline (you will have to pass the paths to all the template.json files in our repo in the yaml file). Same goes for other repos. I imagine that there will be a property where each repo will set in their own pipeline which will point to the root folder of all the templates. Then the tool will go over all the templates under this folder.

Reply:
I replaced the logic here with a bit simpler batching and didn't use `Parallel.ForEach` since it lacks support for async.